veles/secrets/dockerhubpat/validator_test.go

```diff
@@ -0,0 +1,269 @@
+)
+
+const validatorTestPat = "dckr_oat_7awgM4jG5SQvxcvmNzhKj8PQjxo"
+const validatorTestUsername = "User123"
+
+// mockTransport redirects requests to the test server
+type mockTransport struct {
+	testServer *httptest.Server
+}
+
+func (m *mockTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Replace the original URL with our test server URL
+	if req.URL.Host == "hub.docker.com" {
+		testURL, _ := url.Parse(m.testServer.URL)
+		req.URL.Scheme = testURL.Scheme
+		req.URL.Host = testURL.Host
+	}
+	return http.DefaultTransport.RoundTrip(req)
+}
+
+// mockDockerHubServer creates a mock Docker Hub API server for testing
+func mockDockerHubServer(t *testing.T, expectedKey string, expectedUser string) *httptest.Server {
+	t.Helper()
+
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Check if it's a POST request to the expected endpoint
+		if r.Method != http.MethodPost ||
+			r.URL.Path != "/v2/auth/token/" ||
+			r.Header.Get("Content-Type") != "application/json" {
+			t.Errorf("unexpected request: %s %s, expected: POST /v2/auth/token/ with content type application/json", r.Method, r.URL.Path)
+			http.Error(w, "not found", http.StatusNotFound)
+			return
+		}
+
+		// Check Body
+		bodyBytes, err := io.ReadAll(r.Body)
+		if err != nil {
+			t.Errorf("failed to read request body: %v", err)
+			http.Error(w, "internal server error", http.StatusInternalServerError)
+			return
+		}
+		defer r.Body.Close()
+		bodyString := string(bodyBytes)
+		expectedBody := fmt.Sprintf("{\"identifier\": \"%s\",\"secret\": \"%s\"}", expectedUser, expectedKey)
+		if expectedBody != bodyString {
+			w.WriteHeader(http.StatusUnauthorized)
+		}
+		w.WriteHeader(http.StatusOK)
+	}))
+}
+
+func TestValidator(t *testing.T) {
+	cases := []struct {
+		name        string
+		Username    string
+		Pat         string
+		want        veles.ValidationStatus
+		expectError bool
+	}{
+		{
+			name:     "invalid key and username",
+			Pat:      "dckr_oat_random",
+			Username: "User2",
+			want:     veles.ValidationInvalid,
+		},
+		{
+			name:     "valid key and username",
+			Pat:      validatorTestPat,
+			Username: validatorTestUsername,
+			want:     veles.ValidationValid,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// Create a mock server
+			server := mockDockerHubServer(t, validatorTestPat, validatorTestUsername)
+			defer server.Close()
+
+			// Create a client with custom transport
+			client := &http.Client{
+				Transport: &mockTransport{testServer: server},
+			}
+
+			// Create a validator with a mock client
+			validator := dockerhubpat.NewValidator(
+				dockerhubpat.WithClient(client),
+			)
+
+			// Create a test username and pat
+			usernamePat := dockerhubpat.DockerHubPAT{Pat: tc.Pat, Username: tc.Username}
+
+			// Test validation
+			got, err := validator.Validate(context.Background(), usernamePat)
+
+			// Check error expectation
+			if tc.expectError {
```

Comment:
Less verbose to use cmp.Equal with nil or cmpopts.AnyError - see https://github.com/google/osv-scalibr/blob/730d1c644e74ade6a68c9f26ff3ae6aba6a2ef6d/plugin/plugin_test.go#L102as an example